### PR TITLE
:sparkles: Switched to "kustomize edit set image"

### LINKED ISSUE #754: switch to kustomize `images` field from patch
kustomize supports `images` as a first class transformer, so we should use that instead of creating a patch.

/kind feature

### PR DESCRIPTION
**Pull request by hpandeycodeit**
Fixes #754 

Switched to "images" tag as per discussion in #754.

### CODE REVIEW

**Comment by DirectXMan12** [comment]
@Liujingfang1 or @mengqiy ptal

**Comment by mengqiy** [comment]
>The following images will be added to kustomization.yaml:
```
images:
- name: busybox
  newName: alpine
  newTag: 3.6
```

Per [kustomize document](https://github.com/kubernetes-sigs/kustomize/blob/master/examples/image.md), `images` should be in `kustomization.yaml`.
More concretely you need to insert proper `images` field in https://github.com/kubernetes-sigs/kubebuilder/blob/ef5d3c4c47703580cb81d48a737e88ebabb9d5fd/pkg/scaffold/v2/manager/kustomization.go and you can completely remove https://github.com/kubernetes-sigs/kubebuilder/blob/master/pkg/scaffold/v2/kustomize_image_patch.go.
You should not touch v1 scaffolding files e.g. pkg/scaffold/v1/kustomize_image_patch.go, since we don't plan to change how v1 works.

File: pkg/scaffold/v2/manager/kustomization.go [hunk 1]
```diff
@@ -27,13 +27,15 @@ var _ input.File = &Kustomization{}
 // Kustomization scaffolds the Kustomization file in manager folder.
 type Kustomization struct {
 	input.Input
+	Images string
```
Comment (mengqiy) [inline pkg/scaffold/v2/manager/kustomization.go]: Sorry for not being clear enough.
The `images` fields I meant is in the generated `kustomization.yaml`, which should eventually looks like:
```yaml
namespace: project-v2-system
namePrefix: project-v2-

bases:
- ...
patches:
- ...

images:
- name: controller:latest
  newName: your/image
  newTag: yourTag
```

After thinking it more, I feel we don't have to change this file. We can change https://github.com/kubernetes-sigs/kubebuilder/blob/e960cb4216040355c39f85bf4a0febccfe983456/pkg/scaffold/v2/makefile.go#L101, which should be more easier for users to use.

**Comment by hpandeycodeit** [comment]
Don't know why this PR got closed. I have created another PR with the suggested changes : 

https://github.com/kubernetes-sigs/kubebuilder/pull/808

Thanks @mengqiy for the explanation!

**Comment by mengqiy** [comment]
hmm, GitHub says you closed the PR.
Anyway, I will review #808

**Comment by hpandeycodeit** [comment]
I just rebased my branch :( 
sorry about that.

**Comment by mengqiy** [comment]
Aha It seems we can resume it in this PR.

**Comment by mengqiy** [comment]
You are heading to the right direction.
In additional to the change you have, you also need to delete https://github.com/kubernetes-sigs/kubebuilder/blob/master/pkg/scaffold/v2/kustomize_image_patch.go and delete [this line](https://github.com/kubernetes-sigs/kubebuilder/blob/a8b683da3890198fed5cc6873be73ff711862a50/pkg/scaffold/project.go#L184).
In the end, you need to run `./generated_golden.sh` to update the test project.

**Comment by hpandeycodeit** [comment]
Thank you @mengqiy 

I am getting these errors while running `./generated_golden/sh`

```
?   	project/pkg/apis	[no test files]
?   	project/pkg/apis/creatures	[no test files]
2019/06/20 16:05:54 failed to start the controlplane. retried 5 times: fork/exec /usr/local/kubebuilder/bin/etcd: no such file or directory
FAIL	project/pkg/apis/creatures/v2alpha1	0.308s
?   	project/pkg/apis/crew	[no test files]
2019/06/20 16:05:54 failed to start the controlplane. retried 5 times: fork/exec /usr/local/kubebuilder/bin/etcd: no such file or directory
FAIL	project/pkg/apis/crew/v1	0.270s
?   	project/pkg/apis/policy	[no test files]
2019/06/20 16:05:53 failed to start the controlplane. retried 5 times: fork/exec /usr/local/kubebuilder/bin/etcd: no such file or directory
FAIL	project/pkg/apis/policy/v1beta1	0.283s
?   	project/pkg/apis/ship	[no test files]
2019/06/20 16:05:54 failed to start the controlplane. retried 5 times: fork/exec /usr/local/kubebuilder/bin/etcd: no such file or directory
FAIL	project/pkg/apis/ship/v1beta1	0.257s
?   	project/pkg/controller	[no test files]
2019/06/20 16:05:55 failed to start the controlplane. retried 5 times: fork/exec /usr/local/kubebuilder/bin/etcd: no such file or directory
FAIL	project/pkg/controller/firstmate	0.232s
2019/06/20 16:05:56 failed to start the controlplane. retried 5 times: fork/exec /usr/local/kubebuilder/bin/etcd: no such file or directory
FAIL	project/pkg/controller/frigate	0.226s
2019/06/20 16:05:55 failed to start the controlplane. retried 5 times: fork/exec /usr/local/kubebuilder/bin/etcd: no such file or directory
FAIL	project/pkg/controller/healthcheckpolicy	0.240s
2019/06/20 16:05:56 failed to start the controlplane. retried 5 times: fork/exec /usr/local/kubebuilder/bin/etcd: no such file or directory
FAIL	project/pkg/controller/kraken	0.194s
2019/06/20 16:05:59 failed to start the controlplane. retried 5 times: fork/exec /usr/local/kubebuilder/bin/etcd: no such file or directory
FAIL	project/pkg/controller/namespace	0.180s
```

I tried setting `export PATH=$PATH:kubebuilder/bin/`  but no help. 

Is there something I am missing here?

**Comment by mengqiy** [comment]
Can you try to ensure you have `kube-apiserver` and `etcd` in your `PATH`.

**Comment by hpandeycodeit** [comment]
Still getting the same errors even after adding the `kube-apiserver` and `etcd` to the `PATH`
:(

**Comment by mengqiy** [comment]
You can set `KUBEBUILDER_ATTACH_CONTROL_PLANE_OUTPUT` to `true` (e.g. `export KUBEBUILDER_ATTACH_CONTROL_PLANE_OUTPUT=true`) and then run `./generated_golden.sh`. You be able to see the kube-apiserver log and etcd log. You may get an idea why the control plane is not coming up.

**Comment by hpandeycodeit** [comment]
> You can set `KUBEBUILDER_ATTACH_CONTROL_PLANE_OUTPUT` to `true` (e.g. `export KUBEBUILDER_ATTACH_CONTROL_PLANE_OUTPUT=true`) and then run `./generated_golden.sh`. You be able to see the kube-apiserver log and etcd log. You may get an idea why the control plane is not coming up.

Unfortunately, it's not printing the control plane output with the above setting. 
I am trying to build it in my mac. Is there something I need to set/change for mac?

**Comment by jimmidyson** [comment]
`./generated_golden.sh` works for me with your branch. Try adding absolute paths to `PATH`, i.e. `export PATH=$PATH:/full/path/tokubebuilder/bin/`, where `/full/path/tokubebuilder/bin/` contains `etcd` and `kube-apiserver`.

**Comment by jimmidyson** [comment]
From the errors above, check if `/usr/local/kubebuilder/bin/etcd` exists and if it runs properly.

**Comment by hpandeycodeit** [comment]
@jimmidyson @mengqiy I am able to run `./generated_golden.sh` apprently there were 2 `etcd` processes running and that was causing the issue. 

Thanks for the pointers!

File: go.mod [hunk 2]
```diff
@@ -4,17 +4,16 @@ go 1.12
...
 	github.com/spf13/afero v1.2.2
 	github.com/spf13/cobra v0.0.3
 	github.com/spf13/pflag v1.0.3
-	golang.org/x/net v0.0.0-20190514140710-3ec191127204 // indirect
 	golang.org/x/sys v0.0.0-20190506115046-ca7f33d4116e // indirect
 	golang.org/x/text v0.3.2 // indirect
 	golang.org/x/tools v0.0.0-20190506145303-2d16b83fe98c
 	gopkg.in/check.v1 v1.0.0-20180628173108-788fd7840127 // indirect
 	gopkg.in/yaml.v2 v2.2.2
+	sigs.k8s.io/kind v0.4.0 // indirect
```
Comment (mengqiy) [inline go.mod]: Please drop this. I believe this is caused by running `go get sigs.k8s.io/kind` in this repo.
And also please run `go mod tidy` to ensure `go.sum` only have necessary deps.

File: testdata/project-v2/Makefile [hunk 3]
```diff
@@ -54,7 +54,8 @@ generate: controller-gen
 docker-build: test
 	docker build . -t ${IMG}
 	@echo "updating kustomize image patch file for manager resource"
-	sed -i'' -e 's@image: .*@image: '"${IMG}"'@' ./config/default/manager_image_patch.yaml
+	kustomize edit set image $IMG ./config/default/manager_image_patch.yaml
```
Comment (mengqiy) [inline testdata/project-v2/Makefile]: You need to do the following or something better.

- cd config/manager
- kustomize edit set image controller=${IMG}
- cd ../..

Comment (hpandeycodeit) [reply testdata/project-v2/Makefile]: changed it as per above

File: go.mod (hunk 2 above)
Comment (hpandeycodeit) [reply go.mod]: done.

File: pkg/scaffold/v2/makefile.go [hunk 4]
```diff
@@ -98,7 +98,8 @@ generate: controller-gen
 docker-build: test
 	docker build . -t ${IMG}
 	@echo "updating kustomize image patch file for manager resource"
-	sed -i'' -e 's@image: .*@image: '"${IMG}"'@' ./config/default/manager_image_patch.yaml
+	kustomize edit set image $IMG ./config/default/manager_image_patch.yaml
```
Comment (mengqiy) [inline pkg/scaffold/v2/makefile.go]: You forgot to update it here.

File: testdata/project-v2/Makefile [hunk 5]
```diff
@@ -54,7 +54,8 @@ generate: controller-gen
 docker-build: test
 	docker build . -t ${IMG}
 	@echo "updating kustomize image patch file for manager resource"
-	sed -i'' -e 's@image: .*@image: '"${IMG}"'@' ./config/default/manager_image_patch.yaml
+	cd config/manager && kustomize edit set image controller=${IMG} && cd ../..
```
Comment (jimmidyson) [inline testdata/project-v2/Makefile]: Don't need the final `&& cd ../..` - each line in a Makefile is run in its own shell so you don't need to change back to previous directory.

**Comment by jimmidyson** [comment]
Can also remove https://github.com/kubernetes-sigs/kubebuilder/blob/41ed2259ae4cbb744a80033efbacd952bee2ed2d/pkg/scaffold/v2/kustomize.go#L78. Will need to regenerate golden files for testing after that too.

File: pkg/scaffold/v2/makefile.go (hunk 4 above)
Comment (hpandeycodeit) [reply pkg/scaffold/v2/makefile.go]: thanks. Updated it now.

File: testdata/project-v2/Makefile (hunk 5 above)
Comment (hpandeycodeit) [reply testdata/project-v2/Makefile]: changed this as well

File: pkg/scaffold/v2/makefile.go [hunk 6]
```diff
@@ -98,7 +98,7 @@ generate: controller-gen
 docker-build: test
 	docker build . -t ${IMG}
 	@echo "updating kustomize image patch file for manager resource"
-	sed -i'' -e 's@image: .*@image: '"${IMG}"'@' ./config/default/manager_image_patch.yaml
+        cd config/manager && kustomize edit set image controller=${IMG}
```
Comment (mengqiy) [inline pkg/scaffold/v2/makefile.go]: nit: please fix the indentation. We should use "\t" here.

Comment (hpandeycodeit) [reply pkg/scaffold/v2/makefile.go]: I am able to run it fine on my local 

```
cd testdata/project-v2
kustomize build config/default
```

File: pkg/scaffold/v2/makefile.go [hunk 7]
```diff
@@ -98,7 +98,7 @@ generate: controller-gen
 docker-build: test
 	docker build . -t ${IMG}
 	@echo "updating kustomize image patch file for manager resource"
```
Comment (mengqiy) [inline pkg/scaffold/v2/makefile.go]: The following 2 lines should be in the `deploy` target.

Comment (hpandeycodeit) [reply pkg/scaffold/v2/makefile.go]: moved them as suggested. thx

**Comment by mengqiy** [comment]
Maybe I'm not clear enough...
The `deploy` and `docker-build` targets should look like the following in `pkg/scaffold/v2/makefile.go`.
```
deploy: manifests
	kubectl apply -f config/crd/bases
	cd config/manager && kustomize edit set image controller=${IMG}
	kustomize build config/default | kubectl apply -f -
```

```
docker-build: test
	docker build . -t ${IMG}
```

And you will need to run `generated_golden.sh` to sync the test project.

File: testdata/project-v2/Makefile [hunk 8]
```diff
@@ -54,7 +54,7 @@ generate: controller-gen
 docker-build: test
 	docker build . -t ${IMG}
 	@echo "updating kustomize image patch file for manager resource"
-	sed -i'' -e 's@image: .*@image: '"${IMG}"'@' ./config/default/manager_image_patch.yaml
+        cd config/manager && kustomize edit set image controller=${IMG}
```
Comment (jimmidyson) [inline testdata/project-v2/Makefile]: Indentation must be with tabs, not spaces.

**Comment by hpandeycodeit** [comment]
> Maybe I'm not clear enough...
> The `deploy` and `docker-build` targets should look like the following in `pkg/scaffold/v2/makefile.go`.
> 
> ```
> deploy: manifests
> 	kubectl apply -f config/crd/bases
> 	cd config/manager && kustomize edit set image controller=${IMG}
> 	kustomize build config/default | kubectl apply -f -
> ```
> 
> ```
> docker-build: test
> 	docker build . -t ${IMG}
> ```
> 
> And you will need to run `generated_golden.sh` to sync the test project.

My bad. Regenerated the `generated_golden.sh` and pushed the changes again.

File: testdata/project-v2/Makefile (hunk 8 above)
Comment (hpandeycodeit) [reply testdata/project-v2/Makefile]: updated it. thx

**Comment by hpandeycodeit** [comment]
Integration tests are still failing. I tried running `./test_e2e.sh` but I am getting `rbac` errors. How can I run the same test locally?

```

[Fail] kubebuilder with v2 scaffolding [It] should generate a runnable project 
/home/travis/gopath/src/sigs.k8s.io/kubebuilder/test/e2e/e2e_v2.go:127
Ran 2 of 2 Specs in 429.391 seconds
FAIL! -- 1 Passed | 1 Failed | 0 Pending | 0 Skipped
--- FAIL: TestE2E (429.39s)
FAIL
FAIL	sigs.k8s.io/kubebuilder/test/e2e	429.398s
The command "if [ "$TRAVIS_OS_NAME" = "linux" ]; then GO111MODULE=on TRACE=1 PATH=$PATH:$(pwd) ./test_e2e.sh ; fi" exited with 1.

```

**Comment by mengqiy** [comment]
> deploy: manifests
	kubectl apply -f config/crd/bases
	cd config/manager && kustomize edit set image controller=${IMG}
	kustomize build config/default | kubectl apply -f -

`cd config/manager && kustomize edit set image controller=${IMG}` should be `cd config/manager && kustomize edit set image controller=${IMG} && cd ../..` since `kustomize build config/default` is expecting a different work dir.

**Comment by jimmidyson** [comment]
@mengqiy No the `cd ../..` is not needed as every command in a Makefile is run in it's own shell (the working dir is effectively reset).

**Comment by mengqiy** [comment]
> @mengqiy No the cd ../.. is not needed as every command in a Makefile is run in it's own shell (the working dir is effectively reset).

Got it.

**Comment by mengqiy** [comment]
> How can I run the same test locally?

You can also run https://github.com/kubernetes-sigs/kubebuilder/blob/master/scripts/install_and_setup.sh to create a k8s cluster using `kind`. And then running `test_e2e.sh` should be able to reproduce.

**Comment by hpandeycodeit** [comment]
I can run ./test_e2e.sh in local now, however, every time it's failing on different error:

```
   Expected success, but got an error:
        <*errors.errorString | 0xc00018c0f0>: {
            s: "make deploy failed with error: /hp/go/bin/controller-gen \"crd:trivialVersions=true\" rbac:roleName=manager-role webhook paths=\"./...\" output:crd:artifacts:config=config/crd/bases\nkubectl apply -f config/crd/bases\ncustomresourcedefinition.apiextensions.k8s.io/foonhgds.barnhgd.example.comnhgd created\ncd config/manager && kustomize edit set image controller=controller:latest\nkustomize build config/default | kubectl apply -f -\nError: var '{NAMESPACE ~G_v1_Service {metadata.namespace}}' cannot be mapped to a field in the set of known resources\nerror: no objects passed to apply\nmake: *** [deploy] Error 1\n",
        }
        make deploy failed with error: /hp/go/bin/controller-gen "crd:trivialVersions=true" rbac:roleName=manager-role webhook paths="./..." output:crd:artifacts:config=config/crd/bases
        kubectl apply -f config/crd/bases
        customresourcedefinition.apiextensions.k8s.io/foonhgds.barnhgd.example.comnhgd created
        cd config/manager && kustomize edit set image controller=controller:latest
        kustomize build config/default | kubectl apply -f -
        Error: var '{NAMESPACE ~G_v1_Service {metadata.namespace}}' cannot be mapped to a field in the set of known resources
        error: no objects passed to apply
        make: *** [deploy] Error 1
        

```

Or

```

------------------------------
• Failure in Spec Setup (BeforeEach) [4.772 seconds]
kubebuilder
/hp/go/src/sigs.k8s.io/kubebuilder/test/e2e/e2e_v2.go:32
  with v2 scaffolding
  /hp/go/src/sigs.k8s.io/kubebuilder/test/e2e/e2e_v2.go:33
    should generate a runnable project [BeforeEach]
    /hp/go/src/sigs.k8s.io/kubebuilder/test/e2e/e2e_v2.go:56

    Expected success, but got an error:
        <*errors.errorString | 0xc000066670>: {
            s: "kubectl create namespace cert-manager failed with error: Error from server (AlreadyExists): namespaces \"cert-manager\" already exists\n",
        }
        kubectl create namespace cert-manager failed with error: Error from server (AlreadyExists): namespaces "cert-manager" already exists
        
    /hp/go/src/sigs.k8s.io/kubebuilder/test/e2e/e2e_v2.go:42


```

**Comment by mengqiy** [comment]
The 1st error is caused by https://github.com/kubernetes-sigs/kubebuilder/issues/850. I'm going to fix it.
The 2nd error is caused by some leftover resources. e2e tests should be able to cleanup its resources, if you let it run through.

You can recreate the kind cluster:
```
kind delete cluster
kind create cluster --config test/kind-config.yaml --image=kindest/node:v1.14.1
```

**Comment by hpandeycodeit** [comment]
Not getting the 2nd error anymore. But the 1st one still there. And looks like the Travis integration-test still failing at this error.

**Comment by mengqiy** [comment]
@hpandeycodeit The 1st issue should have been fixed on HEAD. You need kustomize **v3.0.1+**.

**Comment by hpandeycodeit** [comment]
@mengqiy I am still getting this error when running e2e in my local:

```
Expected success, but got an error:
        <*errors.errorString | 0xc0000b3880>: {
            s: "make deploy IMG=e2e-test/controller-manager:kvtj failed with error: /Users/hpandey/go/bin//controller-gen \"crd:trivialVersions=true\" rbac:roleName=manager-role webhook paths=\"./...\" output:crd:artifacts:config=config/crd/bases\nkubectl apply -f config/crd/bases\ncustomresourcedefinition.apiextensions.k8s.io/fookvtjs.barkvtj.example.comkvtj created\ncd config/manager && kustomize edit set image controller=e2e-test/controller-manager:kvtj\nkustomize build config/default | kubectl apply -f -\nError: var '{NAMESPACE ~G_v1_Service {metadata.namespace}}' cannot be mapped to a field in the set of known resources\nerror: no objects passed to apply\nmake: *** [deploy] Error 1\n",
        }
        make deploy IMG=e2e-test/controller-manager:kvtj failed with error: /Users/hpandey/go/bin//controller-gen "crd:trivialVersions=true" rbac:roleName=manager-role webhook paths="./..." output:crd:artifacts:config=config/crd/bases
        kubectl apply -f config/crd/bases
        customresourcedefinition.apiextensions.k8s.io/fookvtjs.barkvtj.example.comkvtj created
        cd config/manager && kustomize edit set image controller=e2e-test/controller-manager:kvtj
        kustomize build config/default | kubectl apply -f -
        Error: var '{NAMESPACE ~G_v1_Service {metadata.namespace}}' cannot be mapped to a field in the set of known resources
        error: no objects passed to apply
        make: *** [deploy] Error 1
        

```

**Comment by jimmidyson** [comment]
Try an older version of kustomize, 2.0.3

**Comment by mengqiy** [comment]
I have fixed your PR in #862. Tests are passing. I picked all of your commits and squashed them in the 1st commit and then fix the test in the 2nd commit.
We can either merge #862 or you can copy the fix into your PR.

**Comment by mengqiy** [comment]
https://github.com/kubernetes-sigs/kubebuilder/pull/862/commits/f763d3530492ae8ce71910376e8427e452a0885b is the fix you need to get e2e tests passing

**Comment by hpandeycodeit** [comment]
I am trying with the change in #862 in my local. Let me give one more try. If it doesn't work then you can merge the changes in #862. Could be something specific to my local settings.

**Comment by mengqiy** [comment]
@hpandeycodeit you need to rebase, otherwise CI won't run.

**Comment by mengqiy** [comment]
It's passing now.
Please squash the commits.

**Comment by mengqiy** [comment]
@hpandeycodeit  I guess you mess up something when squashing the commits.
You can find the diff comparing #862.

**Comment by hpandeycodeit** [comment]
> @hpandeycodeit I guess you mess up something when squashing the commits.
> You can find the diff comparing #862.

Thanks! Just added those.

**Comment by hpandeycodeit** [comment]
@k8s-ci-robot /retest

**Comment by mengqiy** [comment]
LGTM
Please squash commits before merging.

**Comment by hpandeycodeit** [comment]
@k8s-ci-robot /retest

**Comment by hpandeycodeit** [comment]
Thank you @mengqiy for the guidance!

**Pull request merged**